feature/guarantor/src/commonMain/kotlin/org/mifos/mobile/feature/guarantor/screens/guarantorDetails/GuarantorDetailViewModel.kt

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2024 Mifos Initiative
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * See https://github.com/openMF/mobile-mobile/blob/master/LICENSE.md
+ */
+package org.mifos.mobile.feature.guarantor.screens.guarantorDetails
+
+import androidx.lifecycle.SavedStateHandle
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import mifos_mobile.feature.guarantor.generated.resources.Res
+import mifos_mobile.feature.guarantor.generated.resources.guarantor_deleted_successfully
+import org.jetbrains.compose.resources.getString
+import org.mifos.mobile.core.common.Constants.INDEX
+import org.mifos.mobile.core.common.Constants.LOAN_ID
+import org.mifos.mobile.core.common.DataState
+import org.mifos.mobile.core.data.repository.GuarantorRepository
+import org.mifos.mobile.core.data.util.NetworkMonitor
+import org.mifos.mobile.core.model.IgnoredOnParcel
+import org.mifos.mobile.core.model.Parcelable
+import org.mifos.mobile.core.model.Parcelize
+import org.mifos.mobile.core.model.entity.guarantor.GuarantorPayload
+import org.mifos.mobile.core.ui.utils.BaseViewModel
+
+/**
+ * Currently we do not get back any response from the guarantorApi, hence we are using FakeRemoteDataSource
+ * to show a list of guarantors. You can look at the implementation of [GuarantorRepository] for better understanding
+ */
+
+internal class GuarantorDetailViewModel(
+    private val guarantorRepositoryImp: GuarantorRepository,
+    savedStateHandle: SavedStateHandle,
+    networkMonitor: NetworkMonitor,
+) : BaseViewModel<GuarantorDetailState, GuarantorDetailEvent, GuarantorDetailAction>(
+    initialState = GuarantorDetailState(
+        dialogState = null,
+        loanId = savedStateHandle.getStateFlow<Long?>(LOAN_ID, null).value,
+        index = savedStateHandle.getStateFlow<Int?>(INDEX, null).value,
+    ),
+) {
+
+    init {
+        viewModelScope.launch {
+            networkMonitor.isOnline.collect { isConnected ->
+                updateState { it.copy(isOnline = isConnected) }
+            }
+        }
+        getGuarantorItem()
+    }
+
+    private fun updateState(update: (GuarantorDetailState) -> GuarantorDetailState) {
+        mutableStateFlow.update(update)
+    }
+
+    private fun getGuarantorItem() {
+        viewModelScope.launch {
+            updateState { it.copy(dialogState = GuarantorDetailState.DialogState.Loading) }
+
+            state.loanId?.let {
+                guarantorRepositoryImp.getGuarantorList(loanId = it)
```

Comment:
@itsPronay One possible reason you might still see the loading bar is that the state is being set to Loading in `getGuarantorItem()`, and later, if loanId is null, the state never gets updated. This means the loading state remains active indefinitely, causing the spinner to keep showing. You may need to handle the case where loanId is null and ensure that the state is properly updated to stop the loading indicator. 

one way to fix it can be doing this code at the end of let block. I don't say it is good way of doing it just mentioning you handle it in better way.
```
state.loanId?.let {
....
} ?: run {
            // If no loanId, stop loading state
            updateState { it.copy(dialogState = null) }
        }
```